File: Go/slug_utils/slug_utils_test.go
package slug_utils

import (
	"testing"
)

func TestGenerate(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected string
	}{
		{"simple", "Hello World", "hello-world"},
		{"with numbers", "Hello World 123", "hello-world-123"},
		{"special chars", "Hello!@#World", "hello-world"},
		{"multiple spaces", "Hello    World", "hello-world"},
		{"leading trailing spaces", "  Hello World  ", "hello-world"},
		{"underscores", "hello_world", "hello-world"},
		{"mixed case", "HeLLo WoRLD", "hello-world"},
		{"empty string", "", ""},
		{"only special chars", "!@#$%^", ""},
		{"with dots", "hello.world", "hello-world"},
		{"with plus", "hello+world", "hello-world"},
		{"consecutive dashes", "hello---world", "hello-world"},
		{"unicode letters", "café", "cafe"},
		{"german umlaut", "grüße", "gruesse"},
		{"cyrillic", "привет мир", "privet-mir"},
		{"greek", "γεια σου", "geia-sou"},
		{"currency", "price €100", "price-eur-100"},
		{"copyright", "© 2024 Company", "c-2024-company"},
		{"trademark", "Product™", "product-tm"},
		{"chinese", "你好世界", "ni-hao-shi-jie"},
		{"japanese", "こんにちは", "konnichiha"},
		{"korean", "안녕하세요", "annyeonghaseyo"},
		{"arabic", "مرحبا", "mrhba"},
	}

	slugger := New()

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := slugger.Generate(tt.input)
			if result != tt.expected {
				t.Errorf("Generate(%q) = %q, want %q", tt.input, result, tt.expected)
			}
		})
	}
}

func TestGenerateWithConfig(t *testing.T) {
	t.Run("custom separator underscore", func(t *testing.T) {
		slugger := NewWithConfig(Config{
			Separator: "_",
			Lowercase: true,
		})
		result := slugger.Generate("Hello World")
		expected := "hello_world"
		if result != expected {
			t.Errorf("Generate with underscore = %q, want %q", result, expected)
		}
	})

	t.Run("no lowercase", func(t *testing.T) {
		slugger := NewWithConfig(Config{
			Separator: "-",
			Lowercase: false,
		})
		result := slugger.Generate("Hello World")
		expected := "Hello-World"
		if result != expected {
			t.Errorf("Generate without lowercase = %q, want %q", result, expected)
		}
	})

	t.Run("max length", func(t *testing.T) {
		slugger := NewWithConfig(Config{
			Separator: "-",
			MaxLength: 10,
			Lowercase: true,
		})
		result := slugger.Generate("Hello Beautiful World")
		// Should truncate at word boundary
		if len(result) > 10 {
			t.Errorf("Generate with max length = %q (len=%d), want len <= 10", result, len(result))
		}
	})

	t.Run("max length exact", func(t *testing.T) {
		slugger := NewWithConfig(Config{
			Separator:     "-",
			MaxLength:     5,
			Lowercase:     true,
			TrimSeparator: true,
		})
		result := slugger.Generate("Hello World")
		if len(result) > 5 {
			t.Errorf("Generate with max length 5 = %q (len=%d)", result, len(result))
		}
	})
}

func TestGenerateUnique(t *testing.T) {
	existing := make(map[string]bool)

	// Simulate existing slugs
	existing["hello-world"] = true
	existing["hello-world-1"] = true

	slugger := New()

	t.Run("first unique slug", func(t *testing.T) {
		result := slugger.GenerateUnique("My New Post", existing)
		expected := "my-new-post"
		if result != expected {
			t.Errorf("GenerateUnique = %q, want %q", result, expected)
		}
	})

	t.Run("duplicate base slug", func(t *testing.T) {
		result := slugger.GenerateUnique("Hello World", existing)
		expected := "hello-world-2"
		if result != expected {
			t.Errorf("GenerateUnique = %q, want %q", result, expected)
		}
	})

	t.Run("another duplicate", func(t *testing.T) {
		existing["hello-world-2"] = true
		result := slugger.GenerateUnique("Hello World", existing)
		expected := "hello-world-3"
		if result != expected {
			t.Errorf("GenerateUnique = %q, want %q", result, expected)
		}
	})
}

func TestGenerateMultiple(t *testing.T) {
	slugger := New()

	result := slugger.GenerateMultiple("Hello", "World", "2024")
	expected := "hello-world-2024"
	if result != expected {
		t.Errorf("GenerateMultiple = %q, want %q", result, expected)
	}
}

func TestIsValidSlug(t *testing.T) {
	tests := []struct {
		slug     string
		expected bool
	}{
		{"hello-world", true},
		{"hello_world", true},
		{"hello123", true},
		{"hello-world-123", true},
		{"Hello-World", true},
		{"", false},
		{"hello world", false},
		{"hello@world", false},
		{"hello.world", false},
		{"hello/world", false},
		{"hello world!", false},
	}

	for _, tt := range tests {
		t.Run(tt.slug, func(t *testing.T) {
			result := IsValidSlug(tt.slug)
			if result != tt.expected {
				t.Errorf("IsValidSlug(%q) = %v, want %v", tt.slug, result, tt.expected)
			}
		})
	}
}

func TestParseSlug(t *testing.T) {
	tests := []struct {
		slug      string
		separator string
		expected  []string
	}{
		{"hello-world", "-", []string{"hello", "world"}},
		{"hello_world_123", "_", []string{"hello", "world", "123"}},
		{"single", "-", []string{"single"}},
		{"", "-", []string{""}},
	}

	for _, tt := range tests {
		t.Run(tt.slug, func(t *testing.T) {
			result := ParseSlug(tt.slug, tt.separator)
			if len(result) != len(tt.expected) {
				t.Errorf("ParseSlug length = %d, want %d", len(result), len(tt.expected))
				return
			}
			for i, v := range result {
				if v != tt.expected[i] {
					t.Errorf("ParseSlug[%d] = %q, want %q", i, v, tt.expected[i])
				}
			}
		})
	}
}

func TestTruncate(t *testing.T) {
	tests := []struct {
		name      string
		slug      string
		maxLength int
		separator string
		expected  string
	}{
		{"no truncation needed", "hello", 10, "-", "hello"},
		{"exact length", "hello", 5, "-", "hello"},
		{"truncate at separator", "hello-world-test", 11, "-", "hello-world"},
		{"truncate short", "hello-world", 5, "-", "hello"},
		{"empty separator", "hello-world", 11, "", "hello-world"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := Truncate(tt.slug, tt.maxLength, tt.separator)
			if result != tt.expected {
				t.Errorf("Truncate = %q, want %q", result, tt.expected)
			}
		})
	}
}

func TestIntToString(t *testing.T) {
	tests := []struct {
		input    int
		expected string
	}{
		{0, "0"},
		{1, "1"},
		{123, "123"},
		{-1, "-1"},
		{-123, "-123"},
		{999999, "999999"},
	}

	for _, tt := range tests {
		t.Run(tt.expected, func(t *testing.T) {
			result := intToString(tt.input)
			if result != tt.expected {
				t.Errorf("intToString(%d) = %q, want %q", tt.input, result, tt.expected)
			}
		})
	}
}

func TestTransliteration(t *testing.T) {
	slugger := New()

	tests := []struct {
		name     string
		input    string
		expected string
	}{
		{"french accents", "à la carte", "a-la-carte"},
		{"spanish tilde", "año nuevo", "ano-nuevo"},
		{"german umlaut", "München", "muenchen"},
		{"german sharp s", "straße", "strasse"},
		{"scandinavian", "Öland åker", "oelandaaker"},
		{"portuguese", "coração", "coracao"},
		{"russian", "Москва", "moskva"},
		{"greek", "Αθήνα", "athina"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := slugger.Generate(tt.input)
			if result != tt.expected {
				t.Errorf("Transliteration %q = %q, want %q", tt.input, result, tt.expected)
			}
		})
	}
}

func TestPackageLevelFunctions(t *testing.T) {
	t.Run("Generate", func(t *testing.T) {
		result := Generate("Hello World")
		expected := "hello-world"
		if result != expected {
			t.Errorf("Generate = %q, want %q", result, expected)
		}
	})

	t.Run("GenerateMultiple", func(t *testing.T) {
		result := GenerateMultiple("Hello", "World")
		expected := "hello-world"
		if result != expected {
			t.Errorf("GenerateMultiple = %q, want %q", result, expected)
		}
	})

	t.Run("GenerateUnique", func(t *testing.T) {
		existing := make(map[string]bool)
		result := GenerateUnique("Hello World", existing)
		expected := "hello-world"
		if result != expected {
			t.Errorf("GenerateUnique = %q, want %q", result, expected)
		}
	})
}

// Benchmark tests
func BenchmarkGenerate(b *testing.B) {
	slugger := New()
	input := "Hello Beautiful World! This is a test @#$ string with some special characters."

	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		slugger.Generate(input)
	}
}

func BenchmarkGenerateShort(b *testing.B) {
	slugger := New()
	input := "hello-world"

	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		slugger.Generate(input)
	}
}

func BenchmarkGenerateUnique(b *testing.B) {
	slugger := New()
	existing := make(map[string]bool)

	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		slugger.GenerateUnique("Hello World", existing)
	}
}

func BenchmarkTransliteration(b *testing.B) {
	slugger := New()
	input := "Привет мир! Grüße aus München 你好世界"

	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		slugger.Generate(input)
	}
}

File: PHP/slug_utils/slug_utils_test.php
<?php
/**
 * Tests for SlugUtils
 * 
 * Run with: php slug_utils_test.php
 * 
 * @package SlugUtils\Tests
 */

require_once __DIR__ . '/slug_utils.php';

use SlugUtils\SlugGenerator;
use SlugUtils\SlugHelper;

class SlugUtilsTest
{
    private int $passed = 0;
    private int $failed = 0;
    private array $errors = [];
    
    /**
     * Assert that two values are equal
     */
    private function assertEquals($expected, $actual, string $message = ''): void
    {
        if ($expected === $actual) {
            $this->passed++;
        } else {
            $this->failed++;
            $this->errors[] = "FAILED: $message\n  Expected: '$expected'\n  Actual: '$actual'";
        }
    }
    
    /**
     * Assert that a condition is true
     */
    private function assertTrue(bool $condition, string $message = ''): void
    {
        if ($condition) {
            $this->passed++;
        } else {
            $this->failed++;
            $this->errors[] = "FAILED: $message (expected true, got false)";
        }
    }
    
    /**
     * Run all tests
     */
    public function run(): void
    {
        echo "===================================\n";
        echo "  SlugUtils Test Suite\n";
        echo "===================================\n\n";
        
        $this->testBasicSlugGeneration();
        $this->testTransliteration();
        $this->testCustomSeparator();
        $this->testMaxLength();
        $this->testUniqueSlugGeneration();
        $this->testSlugValidation();
        $this->testSlugSanitization();
        $this->testSlugConversion();
        $this->testHelperFunctions();
        $this->testEdgeCases();
        $this->testInternationalCharacters();
        
        $this->printResults();
    }
    
    /**
     * Test basic slug generation
     */
    private function testBasicSlugGeneration(): void
    {
        echo "Testing basic slug generation...\n";
        
        $generator = new SlugGenerator();
        
        $this->assertEquals(
            'hello-world',
            $generator->generate('Hello World'),
            'Simple space to dash'
        );
        
        $this->assertEquals(
            'this-is-a-test',
            $generator->generate('This is a test'),
            'Multiple spaces'
        );
        
        $this->assertEquals(
            'hello-world',
            $generator->generate('Hello World!!!'),
            'Remove special characters'
        );
        
        $this->assertEquals(
            'hello-world',
            $generator->generate('  Hello  World  '),
            'Trim spaces'
        );
        
        $this->assertEquals(
            'test-123',
            $generator->generate('Test 123'),
            'Preserve numbers'
        );
        
        $this->assertEquals(
            'test-123-abc',
            $generator->generate('Test 123 ABC'),
            'Mix of letters and numbers'
        );
        
        echo "  Basic tests completed\n\n";
    }
    
    /**
     * Test transliteration
     */
    private function testTransliteration(): void
    {
        echo "Testing transliteration...\n";
        
        $generator = new SlugGenerator();
        
        // French
        $this->assertEquals(
            'cafe-francais',
            $generator->generate('Café Français'),
            'French accents'
        );
        
        // German
        $this->assertEquals(
            'gruße-aus-deutschland',
            $generator->generate('Grüße aus Deutschland'),
            'German umlauts'
        );
        
        // Spanish
        $this->assertEquals(
            'nino-espanol',
            $generator->generate('Niño Español'),
            'Spanish tilde'
        );
        
        // Portuguese
        $this->assertEquals(
            'coracao-do-brasil',
            $generator->generate('Coração do Brasil'),
            'Portuguese accents'
        );
        
        echo "  Transliteration tests completed\n\n";
    }
    
    /**
     * Test custom separator
     */
    private function testCustomSeparator(): void
    {
        echo "Testing custom separator...\n";
        
        $generator = new SlugGenerator(['separator' => '_']);
        
        $this->assertEquals(
            'hello_world',
            $generator->generate('Hello World'),
            'Underscore separator'
        );
        
        $this->assertEquals(
            'this_is_a_test',
            $generator->generate('This is a test'),
            'Multiple words with underscore'
        );
        
        $generator2 = new SlugGenerator(['separator' => '.']);
        $this->assertEquals(
            'hello.world',
            $generator2->generate('Hello World'),
            'Dot separator'
        );
        
        echo "  Custom separator tests completed\n\n";
    }
    
    /**
     * Test max length
     */
    private function testMaxLength(): void
    {
        echo "Testing max length...\n";
        
        $generator = new SlugGenerator(['max_length' => 20]);
        
        $this->assertEquals(
            'this-is-a-very',
            $generator->generate('This is a very long string that should be truncated'),
            'Truncate to max length'
        );
        
        $this->assertEquals(
            'short-string',
            $generator->generate('Short String'),
            'Keep short strings'
        );
        
        echo "  Max length tests completed\n\n";
    }
    
    /**
     * Test unique slug generation
     */
    private function testUniqueSlugGeneration(): void
    {
        echo "Testing unique slug generation...\n";
        
        $generator = new SlugGenerator();
        $existing = ['hello-world', 'hello-world-1', 'hello-world-2'];
        
        $this->assertEquals(
            'hello-world',
            $generator->generateUnique('Hello World', []),
            'First occurrence'
        );
        
        $this->assertEquals(
            'hello-world-3',
            $generator->generateUnique('Hello World', $existing),
            'Skip existing slugs'
        );
        
        $this->assertEquals(
            'new-slug',
            $generator->generateUnique('New Slug', $existing),
            'New unique slug'
        );
        
        echo "  Unique slug tests completed\n\n";
    }
    
    /**
     * Test slug validation
     */
    private function testSlugValidation(): void
    {
        echo "Testing slug validation...\n";
        
        $generator = new SlugGenerator();
        
        $this->assertTrue(
            $generator->isValid('hello-world'),
            'Valid slug with dash'
        );
        
        $this->assertTrue(
            $generator->isValid('test123'),
            'Valid alphanumeric slug'
        );
        
        $this->assertTrue(
            $generator->isValid('a-b-c-d'),
            'Valid multi-part slug'
        );
        
        $this->assertTrue(
            !$generator->isValid(''),
            'Empty slug is invalid'
        );
        
        $this->assertTrue(
            !$generator->isValid('Hello World'),
            'Slug with spaces is invalid'
        );
        
        $this->assertTrue(
            !$generator->isValid('hello_world'),
            'Slug with underscore is invalid (wrong separator)'
        );
        
        echo "  Validation tests completed\n\n";
    }
    
    /**
     * Test slug sanitization
     */
    private function testSlugSanitization(): void
    {
        echo "Testing slug sanitization...\n";
        
        $generator = new SlugGenerator();
        
        $this->assertEquals(
            'hello-world',
            $generator->sanitize('--hello--world--'),
            'Remove extra dashes'
        );
        
        $this->assertEquals(
            'hello-world',
            $generator->sanitize('hello---world'),
            'Replace multiple dashes'
        );
        
        $this->assertEquals(
            'test',
            $generator->sanitize('TEST'),
            'Convert to lowercase'
        );
        
        echo "  Sanitization tests completed\n\n";
    }
    
    /**
     * Test slug conversion methods
     */
    private function testSlugConversion(): void
    {
        echo "Testing slug conversion methods...\n";
        
        $generator = new SlugGenerator();
        
        $this->assertEquals(
            'Hello World',
            $generator->toTitle('hello-world'),
            'Convert to title'
        );
        
        $this->assertEquals(
            'helloWorld',
            $generator->toCamelCase('hello-world'),
            'Convert to camelCase'
        );
        
        $this->assertEquals(
            'hello_world',
            $generator->toSnakeCase('hello-world'),
            'Convert to snake_case'
        );
        
        $this->assertEquals(
            ['hello', 'world'],
            $generator->parse('hello-world'),
            'Parse slug into words'
        );
        
        echo "  Conversion tests completed\n\n";
    }
    
    /**
     * Test helper functions
     */
    private function testHelperFunctions(): void
    {
        echo "Testing helper functions...\n";
        
        $this->assertEquals(
            'quick-slug-test',
            SlugHelper::slug('Quick Slug Test!'),
            'SlugHelper::slug()'
        );
        
        $this->assertEquals(
            'quick_slug_test',
            SlugHelper::underscore('Quick Slug Test!'),
            'SlugHelper::underscore()'
        );
        
        $this->assertEquals(
            'my-file.txt',
            SlugHelper::filename('My File!!!', 'txt'),
            'SlugHelper::filename()'
        );
        
        $this->assertEquals(
            'this-is-a-blog-post-title',
            SlugHelper::titleSlug('This is a Blog Post Title That is Too Long for SEO'),
            'SlugHelper::titleSlug() with max length'
        );
        
        $this->assertEquals(
            'awesome-product-sku123',
            SlugHelper::productSlug('Awesome Product!!!', 'SKU123'),
            'SlugHelper::productSlug()'
        );
        
        $username = SlugHelper::usernameSlug('Admin');
        $this->assertTrue(
            $username !== 'admin',
            'SlugHelper::usernameSlug() avoids reserved words'
        );
        
        $id = SlugHelper::id('user');
        $this->assertTrue(
            strpos($id, 'user-') === 0,
            'SlugHelper::id() with prefix'
        );
        
        echo "  Helper function tests completed\n\n";
    }
    
    /**
     * Test edge cases
     */
    private function testEdgeCases(): void
    {
        echo "Testing edge cases...\n";
        
        $generator = new SlugGenerator();
        
        $this->assertEquals(
            '',
            $generator->generate('!!!'),
            'Only special characters'
        );
        
        $this->assertEquals(
            '123',
            $generator->generate('123'),
            'Only numbers'
        );
        
        $this->assertEquals(
            'a-b-c',
            $generator->generate('a"b"c'),
            'Quotes as separators'
        );
        
        $this->assertEquals(
            'test-2024',
            $generator->generate('test-2024'),
            'Already has dashes'
        );
        
        $this->assertEquals(
            'test',
            $generator->generate('test    test'),
            'Multiple spaces become single dash'
        );
        
        echo "  Edge case tests completed\n\n";
    }
    
    /**
     * Test international characters
     */
    private function testInternationalCharacters(): void
    {
        echo "Testing international characters...\n";
        
        $generator = new SlugGenerator();
        
        // Cyrillic (Russian)
        $this->assertEquals(
            'privet-mir',
            $generator->generate('Привет мир'),
            'Russian Cyrillic'
        );
        
        // Greek
        $this->assertEquals(
            'geia-sou-kosme',
            $generator->generate('Γειά σου κόσμε'),
            'Greek'
        );
        
        // Japanese (basic)
        $slug = $generator->generate('こんにちは');
        $this->assertTrue(
            !empty($slug),
            'Japanese Hiragana transliterates'
        );
        
        // Chinese
        $slug = $generator->generate('你好世界');
        $this->assertTrue(
            !empty($slug),
            'Chinese characters transliterate'
        );
        
        // Arabic
        $slug = $generator->generate('مرحبا');
        $this->assertTrue(
            !empty($slug),
            'Arabic characters transliterate'
        );
        
        echo "  International character tests completed\n\n";
    }
    
    /**
     * Print test results
     */
    private function printResults(): void
    {
        echo "===================================\n";
        echo "  Test Results\n";
        echo "===================================\n";
        echo "  Passed: {$this->passed}\n";
        echo "  Failed: {$this->failed}\n";
        echo "  Total:  " . ($this->passed + $this->failed) . "\n";
        
        if ($this->failed > 0) {
            echo "\n  Errors:\n";
            foreach ($this->errors as $error) {
                echo "  - $error\n";
            }
        }
        
        echo "===================================\n";
        
        if ($this->failed === 0) {
            echo "\n✅ All tests passed!\n";
        } else {
            echo "\n❌ Some tests failed.\n";
        }
    }
}

// Run tests
$test = new SlugUtilsTest();
$test->run();